lib/dataset/dataset.cpp

```diff
@@ -272,53 +273,74 @@ bool Dataset::is_readonly() const noexcept { return m_readonly; }
 typename Masks::holder_type union_or(const Masks &currentMasks,
                                      const Masks &otherMasks) {
   typename Masks::holder_type out;
+  for (const auto &masks : {currentMasks, otherMasks}) {
+    for (const auto &[key, item] : masks) {
+      if (item.dtype() != core::dtype<bool>) {
+        std::string optional_message =
+            " Cannot combine non-boolean mask '" + key + "' in operation";
+        except::throw_mismatch_error(core::dtype<bool>, item.dtype(),
+                                     optional_message);
+      } else if (item.unit() != scipp::units::none) {
+        std::string optional_message =
+            " Cannot combine a unit-specified mask '" + key + "' in operation";
+        except::throw_mismatch_error(scipp::units::none, item.unit(),
+                                     optional_message);
```

Comment:
Won't this raise even if the mask is only in one of the inputs? I don't think there is a reason to abort in that case?

Reply:
Sorry I thought we wanted to filter non-boolean from all masks.

So I updated it to catch only if there are same masks, and either or both of them are non-boolean or unit-specified.
But I just used throw `except::{}Error` instead of `excpet::throw_mismatch_error` to simplify it. Is that okay like this?